### PR TITLE
Apply timezone to Timestamp when using bulkcopy for batch insert

### PR DESCRIPTION
**Pull request by tkyc**
Correctly applies the timezone to Timestamps when inserted using batch insert with bulkcopy.

Fixes https://github.com/microsoft/mssql-jdbc/issues/2271

### CODE REVIEW

**Comment by PriyadharshiniP** [comment]
Hi @tkyc 

I appreciate you making the changes.

I observed that the fix had the following problems:
1. Could you kindly have a look at the applyTimezoneTo method, which appears to use 
`Date date = new Date(**cal**.getTimeInMillis());` rather than `Date date = new Date(**ts.getTime()**); }?
If you have a time value other than today's time in your unit test, **BatchExecutionTest** for **_testValidTimezoneForTimestampBatchInsertWithBulkCopy_**, it will begin to break. The unit test would have seemed to be successful because the timestamp and calendar will often have the same current time value. However, this unit test would begin to fail if you had a different timestamp value.
For example, the unit test failed when I modified the timestamp milliseconds from `long ms = Timestamp.valueOf(todayString).getTime();} to `long ms = 1433338533461L;`. that leads to the failure.

image

2. In the event that the batch insert for bulk copy API falls back to the original batch insert, this fix does not seem to function. An additional conversion occurs when the fallback to no bulk copy API is used, leading to a double conversion for GMT.
To replicate this issue, Use the existing `BulkCopyResultSetCursorTest` for `testMultiplePreparedStatementAndResultSet ` unit test method alone. Connection string should have  "useBulkCopyForBatchInsert=true;sendTemporalDataTypesAsStringForBulkCopy=false;". 
Because bulk copy does not allow IDENTITY columns, using bulk copy would fail and result in inaccurate database values with fallback.
image

**Comment by tkyc** [comment]
@PriyadharshiniP Thanks for the feedback. I'll do more investigating and I'll look into correcting the implementation.

> May I ask if the changes need to be done at SQLServerBulkCopy ?

That was my initial thought also, but when I looked over SQLServerBulkCopy it doesn't look like I'm able to apply the timezone in the TDS request for bulkcopy.

For example, I sort of expected code like this in SQLServerBulkCopy, which is part of the normal and regular batch insert code path. The code below takes into consideration the the timezone by sending the timezone info in the TDS request. So the timezone is applied during insertion by the server:

```
                                    tdsWriter.writeEncryptedRPCDateTime2(name,
                                            timestampNormalizedCalendar(calendar, javaType, conn.baseYear()),
                                            subSecondNanos, (valueLength), isOutParam, statement);
```

But I'm guessing because we're doing a bulkcopy, it has to send the timestamp value as is. And so, either the user or driver needs to apply the timezone conversion beforehand.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java [hunk 1]
```diff
@@ -54,6 +60,8 @@ public class BatchExecutionTest extends AbstractTest {
     private static String ctstable3;
     private static String ctstable4;
     private static String ctstable3Procedure1;
+    private static String timestampTable = AbstractSQLGenerator
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:63]: this table should be added to dropTable() for cleanup

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java [hunk 2]
```diff
@@ -99,6 +107,55 @@ public void testBatchUpdateCountTrueOnFirstPstmtSpPrepare() throws Exception {
...
+
+        try (Connection con = DriverManager.getConnection(connectionString); Statement stmt = con.createStatement();
+                PreparedStatement pstmt = con.prepareStatement("INSERT INTO " + timestampTable + " VALUES(?)")) {
+
+            String dropSql = "IF EXISTS (SELECT * FROM dbo.sysobjects WHERE id = object_id(N'"
+                    + TestUtils.escapeSingleQuotes(timestampTable)
+                    + "') and OBJECTPROPERTY(id, N'IsUserTable') = 1) DROP TABLE " + timestampTable;
+            stmt.execute(dropSql);
+
+            String createSql = "CREATE TABLE" + timestampTable + " (c1 DATETIME2(3))";
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:123]: this table should be added to dropTable() or add finally to cleanup

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java [hunk 3]
```diff
@@ -99,6 +107,55 @@ public void testBatchUpdateCountTrueOnFirstPstmtSpPrepare() throws Exception {
...
 
+    @Test
+    public void testValidTimezoneForTimestampBatchInsertWithBulkCopy() throws Exception {
+        Calendar gmtCal = Calendar.getInstance(TimeZone.getTimeZone("GMT"));
+        long ms = 1578743412000L;
+
+        try (Connection con = DriverManager.getConnection(connectionString); Statement stmt = con.createStatement();
+                PreparedStatement pstmt = con.prepareStatement("INSERT INTO " + timestampTable + " VALUES(?)")) {
+
+            String dropSql = "IF EXISTS (SELECT * FROM dbo.sysobjects WHERE id = object_id(N'"
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:118]: dropTableIfExists

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java [hunk 4]
```diff
@@ -13,11 +13,17 @@
...
 import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Statement;
+import java.sql.Timestamp;
+import java.text.SimpleDateFormat;
 import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.TimeZone;
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:26]: pls cleanup unused

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java [hunk 5]
```diff
@@ -99,6 +107,55 @@ public void testBatchUpdateCountTrueOnFirstPstmtSpPrepare() throws Exception {
...
+
+            Timestamp ts0; // Timestamp batch inserted without bulkcopy
+            Timestamp ts1; // Timestamp batch inserted with bulkcopy
+
+            rs.next();
+            ts0 = rs.getTimestamp(1);
+            rs.next();
+            ts1 = rs.getTimestamp(1);
+
+            assertEquals(ts0, ts1);
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:155]: maybe add a bit more debug info if it fails? I think assertEquals will just tell u false it doesn't equal but not the values

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java [hunk 6]
```diff
@@ -99,6 +107,55 @@ public void testBatchUpdateCountTrueOnFirstPstmtSpPrepare() throws Exception {
...
+
+            Timestamp timestamp = new Timestamp(ms);
+
+            pstmt.setTimestamp(1, timestamp, gmtCal);
+            pstmt.addBatch();
+            pstmt.executeBatch();
+        }
+
+        try (Connection con = DriverManager.getConnection(connectionString); Statement stmt = con.createStatement()) {
+            ResultSet rs = stmt.executeQuery("SELECT * FROM " + timestampTable);
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:145]: ResultSet is leaked

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java [hunk 7]
```diff
@@ -2938,6 +2947,44 @@ private boolean checkSQLLength(int length) {
         return true;
     }
 
+    private String getDateFormatPattern(int nanos) {
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java:2950]: add comments on what this does

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java [hunk 8]
```diff
@@ -2938,6 +2947,44 @@ private boolean checkSQLLength(int length) {
...
+            if (i == 0) {
+                pattern.append(".S");
+            }
+            pattern.append("S");
+        }
+
+        return pattern.toString();
+    }
+
+    private java.sql.Timestamp applyTimezoneTo(java.sql.Timestamp ts, Calendar cal) {
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java:2977]: add comment for method

**Comment by PriyadharshiniP** [comment]
> 1. **BatchExecutionTest**

Thanks for fixing the issue.
1.  Additionally, it appears that not all dates—particularly those with daylight savings time—cannot be converted successfully using the applyTimezoneTo method . In your BatchExecutionTest unit test case , if you change the ms to ` long ms = 1696127400000L;` that equates to :

```
GMT: Sunday, October 1, 2023 2:30:00 AM
My local time zone: Sunday, October 1, 2023 1:30:00 PM [GMT+11:00]
```

This time "GMT: Sunday, October 1, 2023 2:30:00 AM" doesn't exist in my local time zone. The non bulk version correctly inserts the date as `2023-10-01 02:30:00.000` but the bulk version incorrectly inserts as `2023-10-01 03:30:00.000` and the BatchExecutionTest  will fail if we use  ` ts0 = rs.getTimestamp(1, gmtCal);. and ts1 = rs.getTimestamp(1, gmtCal);`
I think this is because of using Timestamp.valueOf or any timestamp related methods that will use only local time for conversions.

image

2. As the standard batch insert without bulk copy seems to work in all cases, I started to look re-use the tdsWriter.writeEncryptedRPCDateTime2 in SQLServerBulkCopy itself, however while attempting to fix it, I encounter certain problems.
To store the parameters needed for the writeEncryptedRPCDateTime2 method, I made a new class. Next, update the SQLServerBulkBatchInsertRecord class with those values. Fill in the datetime2 field in SQLServerBulkCopy in accordance with the screenshots below :
image
image
However, the following error appears:  `java.sql.BatchUpdateException: While reading current row from host, a premature end-of-message was encountered--an incoming data stream was interrupted when the server expected to see more data. The host program may have terminated. Ensure that you are using a supported client application programming interface (API).` 
Kindly inform me if you can assist with this.

Could you see if we can accomplish something along these lines?

**Comment by tkyc** [comment]
@PriyadharshiniP Thanks for collaborating.

> As the standard batch insert without bulk copy seems to work in all cases, I started to look re-use the tdsWriter.writeEncryptedRPCDateTime2 in SQLServerBulkCopy

Yeah, I struggled with getting that working as well when I initially looked at the code. In `SQLServerBulkCopy`, it seems the only thing written to the TDS request is column metadata info and data type info and it doesn't look like any input values are written eg. the timestamp in our case for batch insert with bulkcopy. When I traced the batch insert bulkcopy call path:

```
writeToServer -> sendBulkLoadBCP -> ... -> doInsertBulk -> sendBulkCopyCommand
```

In the above call path, I expected to end up somewhere where the timestamp value is written to the TDS request (something/somewhere where the logic `writeEncryptedRPCDateTime2 ` can be used) but I don't. I'll do some more debugging, maybe I missed some things initially.

> Additionally, it appears that not all dates—particularly those with daylight savings time—cannot be converted successfully using the applyTimezoneTo method

Thanks for the help on testing. I'll see what I can do here...

**Comment by tkyc** [comment]
@PriyadharshiniP What's your exact local time zone? I've switched my PC's time zone setting to various different time zones that doesn't have day light savings and wasn't able to repro the issue. I double checked the local time zone used in my JDK and it matches my PC time zone.

**Comment by PriyadharshiniP** [comment]
My local timezone is Australia/Sydney

**Comment by PriyadharshiniP** [comment]
> @PriyadharshiniP What's your exact local time zone? I've switched my PC's time zone setting to various different time zones that doesn't have day light savings and wasn't able to repro the issue with `1696127400000L`. I double checked the local time zone used in my JDK and it matches the PC time zone set when I was experimenting.
> 
> Maybe I need to do something like below explicitly to repro your issue?
> 
> ```
> TimeZone tz = TimeZone.getTimeZone("your-time-zone");
> TimeZone.setDefault(tz);
> ```

I guess you could do something like this in your unit test  :`System.setProperty("user.timezone","Australia/Sydney");`

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java (hunk 5 above)
Comment (tkyc) [reply src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:211]: `assertEquals` will output an expected vs. actual message, so no need for an actual failure message.

**Comment by tkyc** [comment]
@PriyadharshiniP I've just pushed a new fix implementation. I did some testing of it and it seem to have addressed all your concerns so far. Let me know if there's any additional problems. Thanks again for the collab.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java [hunk 9]
```diff
@@ -2480,10 +2480,13 @@ else if (4 >= bulkScale)
...
-                                String timeStampValue = colValue.toString();
-                                tdsWriter.writeTime(java.sql.Timestamp.valueOf(timeStampValue), bulkScale);
+                                Timestamp ts = java.sql.Timestamp.valueOf(colValue.toString());
+                                tdsWriter.writeTime(ts, bulkScale, cal);
                                 // Send only the date part
-                                tdsWriter.writeDate(timeStampValue.substring(0, timeStampValue.lastIndexOf(' ')));
+                                tdsWriter.writeDate(ts.getTime(), cal);
+                                //tdsWriter.writeRPCDateTime2(name,
+                                //        timestampNormalizedCalendar(calendar, javaType, conn.baseYear()),
+                                //        subSecondNanos, TDS.MAX_FRACTIONAL_SECONDS_SCALE, isOutParam);
```
Comment (Jeffery-Wasty) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java:2489]: Remove commented out code

File: src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java [hunk 10]
```diff
@@ -3836,6 +3850,22 @@ void writeTime(java.sql.Timestamp value, int scale) throws SQLServerException {
         writeScaledTemporal(calendar, subSecondNanos, scale, SSType.TIME);
     }
 
+    void writeTime(java.sql.Timestamp value, int scale, Calendar cal) throws SQLServerException {
+        GregorianCalendar calendar = initializeCalender(TimeZone.getDefault());
+        long utcMillis; // Value to which the calendar is to be set (in milliseconds 1/1/1970 00:00:00 GMT)
+        int subSecondNanos;
+        utcMillis = value.getTime();
+        subSecondNanos = value.getNanos();
```
Comment (Jeffery-Wasty) [inline src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java:3858]: Declarations and assignments can be combined.
```suggestion
        long utcMillis = value.getTime(); // Value to which the calendar is to be set (in milliseconds 1/1/1970 00:00:00 GMT)
        int subSecondNanos = value.getNanos();
```

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java [hunk 11]
```diff
@@ -54,6 +60,10 @@ public class BatchExecutionTest extends AbstractTest {
     private static String ctstable3;
     private static String ctstable4;
     private static String ctstable3Procedure1;
+    private static String timestampTable1 = AbstractSQLGenerator
+            .escapeIdentifier(RandomUtil.getIdentifier("timestamptable1"));
+    private static String timestampTable2 = AbstractSQLGenerator
+            .escapeIdentifier(RandomUtil.getIdentifier("timestampTableBatchInsert2"));
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:66]: stupid nitpick, usually we use same identifier as the name.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java [hunk 12]
```diff
@@ -99,6 +109,186 @@ public void testBatchUpdateCountTrueOnFirstPstmtSpPrepare() throws Exception {
...
+            Timestamp timestamp = new Timestamp(ms);
+
+            pstmt.setTimestamp(1, timestamp, gmtCal);
+            pstmt.addBatch();
+            pstmt.executeBatch();
+        }
+
+        // Compare Timestamp values inserted, should be the same
+        try (Connection con = DriverManager.getConnection(connectionString); Statement stmt = con.createStatement()) {
+            try (ResultSet rs = stmt.executeQuery("SELECT * FROM " + timestampTable1)) {
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:146]: why not combine to same try?

**Comment by PriyadharshiniP** [comment]
> @PriyadharshiniP I've just pushed a new fix implementation. I did some testing of it and it seem to have addressed all your concerns so far. Let me know if there's any additional problems. Thanks again for the collab.

Hi @tkyc 
Thanks for fixing it so quickly . 
I checked all dates this year in my local time zone, and it seems that there will still be problems with the bulk copy inserting the date in my local environment around spring forward dates. The time stamp millisecond values are 1712415600000L and 1743935400000L, to be more precise.
Noted the use of `java.sql.Timestamp.valueOf(colValue.toString())` in SQLServerBulkCopy, this seems to be the cause of this issue as it seems to alter the timestamp milliseconds value. Using the `(Timestamp)colValue` seems to work for all dates. Could you please check and fix this as well?

image
image

**Comment by tkyc** [comment]
@PriyadharshiniP Oops, my mistake. Calling valueOf will cause a conversion. Fixed.

**Review by Jeffery-Wasty — DISMISSED** [review]

**Review by lilgreenbird — APPROVED** [review]

**Review by Jeffery-Wasty — APPROVED** [review]

**Comment by PriyadharshiniP** [comment]
> @PriyadharshiniP Oops, my mistake. Calling valueOf will cause a conversion. Fixed.

Thanks @tkyc :
It seems to be fine for my tests.
Please let me  know when the release would be available ?

Regards

**Comment by tkyc** [comment]
@PriyadharshiniP this will make it in the release that's on the 31st of this month.

**Pull request merged**